master/internal/task/allocation.go

```diff
@@ -0,0 +1,330 @@
+	cproto "github.com/determined-ai/determined/master/pkg/container"
+	"github.com/determined-ai/determined/master/pkg/model"
+	"github.com/determined-ai/determined/master/pkg/ptrs"
+)
+
+type (
+	// Allocation encapsulates all the state of a single allocation
+	Allocation struct {
+		// system dependencies
+		db db.DB
+
+		// the persisted representation.
+		model model.Allocation
+
+		// the spec used to start reservations
+		spec TaskSpecMaker
+		// the keys for SSH access to the task.
+		keys *ssh.PrivateAndPublicKeys
+		// The existence of allocations signifies the trial has been allocated.
+		reservations map[cproto.ID]sproto.Reservation
+		// The following fields tracks containers and their states.
+		containers           map[cproto.ID]cproto.Container
+		terminatedFirst      *cproto.ID
+		terminatedContainers map[cproto.ID]sproto.TaskContainerStopped
+		// preemption encapsulates the preemption state of the currently allocated task.
+		// If there is no current task, or it is unallocated, it is nil.
+		preemption Preemption
+		// rendezvous encapsulates logic of rendezvousing containers of the currently
+		// allocated task. If there is no current task, or it is unallocated, it is nil.
+		rendezvous Rendezvous
+		// killed marks that we have intentionally killed the trial, so we can know to ignore
+		// any errors from containers dying.
+		killed bool
+		// we send a kill when we terminate a task forcibly. we terminate forcibly when a container
+		// exits non zero. we don't need to send all these kills, so this exists.
+		killCooldown *time.Time
+	}
+
+	// AllocationExited summarizes the exit status of an allocation.
+	AllocationExited struct {
+		// userRequestedStop when a container unexpectedly exits with 0.
+		UserRequestedStop bool
+		Err               error
+	}
+)
+
+const killCooldown = 30 * time.Second
+
+// NewAllocation returns a new allocation, which tracks allocation state in a fairly generic way.
+func NewAllocation(
+	ctx *actor.Context, taskID model.TaskID, req sproto.AllocateRequest,
+	reservations []sproto.Reservation, spec TaskSpecMaker, keys *ssh.PrivateAndPublicKeys, db db.DB,
+) (*Allocation, error) {
+	containerIDToReservation := map[cproto.ID]sproto.Reservation{}
+	for _, a := range reservations {
+		containerIDToReservation[a.Summary().ID] = a
+	}
+	a := &Allocation{
+		db: db,
+
+		model: model.Allocation{
+			TaskID:       taskID,
+			AllocationID: req.AllocationID,
+			ResourcePool: req.ResourcePool,
+			StartTime:    time.Now().UTC(),
+		},
+
+		reservations:         containerIDToReservation,
+		spec:                 spec,
+		keys:                 keys,
+		preemption:           NewPreemption(req.AllocationID),
+		rendezvous:           NewRendezvous(req.AllocationID, ranksFromReservations(reservations)),
+		containers:           make(map[cproto.ID]cproto.Container),
+		terminatedContainers: make(map[cproto.ID]sproto.TaskContainerStopped),
+	}
+
+	if err := a.db.AddAllocation(&a.model); err != nil {
```

Comment:
The following code can be put in a separate function `ResourceAllocated` since it will be the reverse part of `Close(...)`.

Reply:
sgtm. tbh, i was considering just making this an actor.. it would not be that difficult.